### PR TITLE
Fixes frazil heatflux diagnostic when remapped

### PR DESCRIPTION
**Pull request by ashao**
This addresses #641. I'm pretty sure I set the averaging windows correctly, but could I get another set of eyes to double check? (I don't want to have accidentally disabled all the thermodynamic diagnostics!)

The frazil heatflux was not being conserved under remapping for two reasons
-The v_extensive flag needed to be set to true
-make_frazil is called multiple timss, but only the accumulated tendency is
posted at the end despite layer thicknesses changing in between calls. To
fix this, enable/disable averaging calls were added to organize the
diabatic algorithm into the following sections:
Time0 to Time0 + 0.5*dt: First call to make_frazil
Time0 to Time0 + dt: Everything but make_frazil
Time0+0.5 to Time0 + dt: Last call to make_frazil

### CODE REVIEW

**Comment by ashao** [comment]
Also, do I get the distinction of submitting the commit from "sea water" (I'm on the ferry from Anacortes, WA to Sidney, B.C. currently 48.5386813N, 123.0182250W).

**Comment by ashao** [comment]
Thanks for the helpful comments.

In OM4_05, I checked that the sum of the frazil heatflux in both the native
and remapped z coordinate was the same. For completeness, I'll do your
suggested check and other sanity checks in OM4_025 today.

I see what you're saying about the timestep. I'll sit with the arithmetic
to make sure that the diagnostic returns the heat flux averaged over the
entire thermodynamic timestep while being posted in two half averaging
intervals.

I'll add a comment later today once I've dealt with both issues.

**Comment by adcroft** [comment]
Test started at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/pipelines/2801

**Comment by StephenGriffies** [comment]
@ashao please point me to the directory where you have the diagnostics saved for OM4_05 or OM4_025.  I wish to be sure the 3d heat budget closes with the new frazil diagnostic code.  

Please be sure the diag_table has "months" changed to "days" so that you will generate some output.  

I will check on Tuesday...so no hurry.

**Comment by ashao** [comment]
@StephenGriffies , I've convinced myself that you're right and that we should use the full DT_THERM to calculate the frazil tendency, but to split the averaging interval into two to get the correct z* tendency remapping.

I've included 2 days of OM4_025 output at /work/aes/OM4_025/frazil_diag

**Comment by ashao** [comment]
Just checked the sum of frazil_heat_tendency and hfsifrazil and the two are off by a factor of 2. I'll delve into this more.

**Comment by ashao** [comment]
So as it turns out the initial 0.5*dt_therm is the correct way to calculate the frazil heat flux so that the average over dt_therm is correct.

Define A1 and A2 as the change in heat for a cell which, for simplicity, assume does not change thickness. The diagnostic manager calculates averages as a weighted mean over the requested averaging interval (in this case dt_therm). In the case of two averaging intervals of equal length the weight for each term is 0.5*dt_therm/dt_therm = 0.5. The equation for the weighted-average tendency is then [A1/(0.5*dt_therm)]*0.5 + [A2/(0.5*dt_therm]*0.5} which simplifies to the expected tendency (A1+A2)/dt_therm.

For the sake of transparency I've reverted the previous commit instead of just rewriting the git history.

**Comment by ashao** [comment]
Could you clarify if the output that is stored at /work/aes/OM4_025/frazil_diag no longer allows you to close the heat budget? In 19000101.ocean_month.nc, the frazil_heat_tendency_2d is equal to hfsifrazil and matches the vertical sum of the remapped frazil_heat_tendency in 19000101.ocean_month_z.nc to within ~0.0002W.

One change in this PR that I wanted to emphasize is that the original code accumulated the change in heat from both calls to make_frazil, calculated the flux by dividing by dt_therm, and made one call to post_data where the averaging window was the entire length of the thermodynamic timestep. Now within the thermodynamic timestep, post_data for frazil_heat_tendency is called twice (after each make_frazil call) where the averaging window is half of the thermodynamic timestep which is also the same dt used to calculate a flux from the change in heat. Does that help clarify the origin of the 0.5 factor?

**Comment by StephenGriffies** [comment]
I verified that the frazil term now is consistent across the native and z-regridded output.  Thanks for your help.  

Unfortunately, the 3d heat budget still does not close on the z-grid.  The problem is in the upper 100m of the ocean.  I believe the fault is with the diagnostic boundary_forcing_heat_tendency and boundary_forcing_salt_tendency.  I looked at that code and am very puzzled with what it is doing. I will post a new issue.

So in short, I think you nailed the Frazil bug.  At least the frazil diagnostic is now self-consistent.  Thanks @ashao

**Comment by ashao** [comment]
I had been able to get the 3D salt budget to close with the remapped
diagnostics, so I'm not sure that the boundary_forcing_heat/salt_tendency
terms are the culprit. Let me dig up the script and verify that salt still
does close.

**Comment by Hallberg-NOAA** [comment]
This PR changed the available_diags files for some of the test cases, but they were not updated in MOM_examples.  Whenever a code change alters the output of out test cases, please be sure to update that output at the same time as we accept a PR.  Thank you.

**Comment by ashao** [comment]
@Hallberg-NOAA : Could you clarify if you had detected that some diagnostics had been added/removed? This PR should only have changed cell_methods for frazil_heat_tendency diagnostics.

Also you raise some interesting questions about PR procedures. If a PR is answer-changing, we seem to have been doing a pretty good job of elucidating that in the PR comments. However, I have been pretty lax about mentioning if any of the other runtime-generated output changes. This inherently puts the burden on the person who accepts the pull-request to detect these changes and update MOM6-examples. Short of requiring that any code-contributor runs their own regression tests and always pairs a PR to MOM6 with one to MOM6-examples, perhaps the easiest way forward so that this responsibility is shared would be to require three bullets at the end of each PR:
- Changes ocean.stats: True/False
- Changes available_diags: True/False
- Changes MOM_parameter_doc.all: True/False

**Comment by kshedstrom** [comment]
Some of us are only doing PRs for MOM6, not MOM6-examples, in part because
I'm not running all of MOM6-examples. I haven't been changing
available_diags, though.

**Comment by StephenGriffies** [comment]
This is my fault, as I approved the pull request. 

But my assumption was that if all tests passed, then that meant all is fine,and ready to merge. Apparently this assumption is wrong. In which case the testing process needs further automation. 

Or at least I need to be properly scolded and told what to do before hitting the green buttons... @Alistair please advise...
